scripts/eslint/rules/mark-jsdoc-types-used.cjs

```diff
@@ -0,0 +1,102 @@
+const { TSESTree, ESLintUtils } = require("@typescript-eslint/utils");
+const { createRule } = require("./utils.cjs");
+const ts = require("typescript");
+
+/**
+ * @param {ts.Identifier} node
+ */
+function isBareIdentifier(node) {
+    if (ts.isPropertyAccessExpression(node.parent)) {
+        return node.parent.expression === node;
+    }
+
+    if (ts.isQualifiedName(node.parent)) {
+        return node.parent.left === node;
+    }
+
+    // TODO(jakebailey): surely I'm missing something here.
+    // Also, how does eslint deal with type space versus value space? Or does it not?
+
+    return true;
+}
+
+module.exports = createRule({
+    name: "mark-jsdoc-types-used",
+    meta: {
+        docs: {
+            description: ``,
+        },
+        messages: {},
+        schema: [],
+        type: "problem",
+        fixable: "whitespace",
+    },
+    defaultOptions: [],
+
+    create(context) {
+        /** @type {(node: TSESTree.Node) => void} */
+        const checkProgram = (node) => {
+            const parserServices = ESLintUtils.getParserServices(context, /*allowWithoutFullTypeInformation*/ true);
+            const ast = parserServices.esTreeNodeToTSNodeMap.get(node);
+
+            // TODO(jakebailey): I almost guarantee this is overzealous and wrong in some way.
+
+            /**
+             * @param {ts.Node} node
+             */
+            function markIdentifiersUsed(node) {
+                if (ts.isIdentifier(node) && isBareIdentifier(node)) {
+                    context.markVariableAsUsed(node.text);
+                    return;
+                }
+                ts.forEachChild(node, markIdentifiersUsed);

```

Comment:
Idk exactly how this traverses - but wouldn't this mark both x and y as used in `x.y`?

Reply:
It shouldn't; `markIdentifiersUsed` checks for that case and only marks things on the left side.